samples: nrf9160: add REST A-GPS API sample (#5755)
nrfconnect/sdk-nrf: samples/nrf9160/nrf_cloud_rest_agps/README.rst

Note that the radio component on the nRF9160-based device is shared between LTE and GPS and hence only one of these can be active at any moment of time.
LTE has priority over the radio resources, and GPS operates in the time interval between LTE required operations as per LTE specification.
Consequently, GPS is blocked from searching as long as the modem is in `RRC connected mode <RRC idle mode_>`_.

The modem is in RRC connected mode when sending and receiving data over the LTE network.
After data transmission is complete, the modem stays in RRC connected mode for a period of network-specific inactivity time, before entering RRC idle mode.
The inactivity time varies between networks.
It is typically in the range of 5 to 70 seconds for a specific network.
In RRC idle mode, the GPS is usually able to operate.

It is recommended to use LTE Power Saving Mode (PSM) and extended Discontinuous Reception (eDRX) mode to increase the allowed time of operation for GPS.
During the defined intervals of PSM and eDRX, LTE communication does not occur, and the GPS has full access to the radio resources.
In this sample, both PSM and eDRX are enabled by default.

> You can enable or disable these features by using the :kconfig:`CONFIG_LTE_POWER_SAVING_MODE` and :kconfig:`CONFIG_LTE_EDRX_REQ` configuration options.

Author: I'm confused by your suggestion.  I do not find any other example anywhere in the sdk-nrf repository where :option: is used in an .rst file.

> You can enable or disable these features by using the :kconfig:`CONFIG_LTE_POWER_SAVING_MODE` and :kconfig:`CONFIG_LTE_EDRX_REQ` configuration options.

Reviewer: This is a fairly new thing to myself as well. As this option is not introduced in Kconfig, we need to refer to the definition given in the same document a bit below. @umapraseeda, please confirm I have understood this correctly.

> You can enable or disable these features by using the :kconfig:`CONFIG_LTE_POWER_SAVING_MODE` and :kconfig:`CONFIG_LTE_EDRX_REQ` configuration options.

Reviewer: So, this option is not included in the Kconfig file at all. If you decide to add it, this can remain as it was :kconfig:, but due to some recent changes in the doc framework, you would need to change the line 90. I'll make a suggestion accordingly on that line.

> You can enable or disable these features by using the :kconfig:`CONFIG_LTE_POWER_SAVING_MODE` and :kconfig:`CONFIG_LTE_EDRX_REQ` configuration options.

Author: I copied the readme from another sample and edited it.  I hadn't noticed that the CONFIG_LTE_POWER_SAVING_MODE option was missing from my sample.

*************
|config|

Configuration options
=====================

Check and configure the following configuration options for the sample:

.. option:: CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode

This configuration option enables or disables the LTE Power Saving Mode.

Additional configuration
========================

Check and configure the following library options that are used by the sample:

* :kconfig:`CONFIG_LTE_EDRX_REQ`

Building and running
********************

.. |sample path| replace:: :file:`samples/nrf9160/nrf_cloud_rest_agps`

.. include:: /includes/build_and_run_nrf9160.txt

The configuration file for this sample is located in :file:`samples/nrf9160/nrf_cloud_rest_agps`.
See :ref:`configure_application` for information on how to configure the parameters.

Testing
=======

|test_sample|

#. Place the nRF9160-based device in a location with direct line of sight to a large portion of the sky.
   This is to achieve sufficient satellite coverage to get a position fix.
#. Optionally, connect the nRF9160-based device to a PC with a USB cable. The kit is assigned a COM port (Windows) or ttyACM device (Linux), which is visible in the Device Manager.

> #. Optionally, connect to the kit with a terminal emulator (for example, PuTTY) to see log output. See How to connect with PuTTY for the required settings.

Reviewer: ```suggestion
#. Optionally, connect to the kit with a terminal emulator (for example, PuTTY) to see log output. See :ref:`How to connect with PuTTY  <putty>` for the required settings.
```

#. Log into your nRF Cloud account and select your device.
#. Power on the nRF9160-based device.

   .. note::
      The sample outputs shown in the subsequent steps correspond to the scenario when nRF Cloud is used as the A-GPS source.

> #. The device connects to the LTE network and nRF Cloud as shown below:

Reviewer: ```suggestion
#. The device connects to the LTE network and nRF Cloud as follows:
```

   .. code-block:: console

	  -------------------------
	  I: A-GPS sample has started
	  I: Connecting to LTE network. This may take minutes.
	  I: PSM mode requested
	  +CEREG: 2,"7725","0138E000",7,0,0,"11100000","11100000"
	  +CSCON: 1
	  +CEREG: 1,"7725","0138E000",7,,,"00000010","00000110"
	  I: Network registration status: Connected - home network

   .. note::
      If the credentials on the device are used to connect to nRF Cloud for the first time, the device will be disconnected, and a :cpp:enumerator:`CLOUD_EVT_DISCONNECTED <cloud_api::CLOUD_EVT_DISCONNECTED>` event will be logged on the terminal.
      This happens due to the configuring of the device on the cloud by `AWS IoT Just-in-time-provisioning`_ and the subsequent termination of the connection.
      In general, the devices reconnect automatically.
      However, for the simplicity of the sample, reconnection is not implemented.

>       Hence you must reset the device manually.

Reviewer: ```suggestion
      Hence, you must reset the device manually.
```

> #. Make sure that GPS has started by confirming that the device requests GPS assistance data and receives a response.

Reviewer: ```suggestion
#. To make sure that GPS has started, confirm that the device requests GPS assistance data and receives a response.
```

>    You can do this by checking the logs.

Reviewer: ```suggestion
   Check the response in the logs.
```

	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  D: Sent A-GPS data to modem, type: 2
	  I: A-GPS data successfully processed
	  I: GPS will start in 3 seconds
	  D: GPS mode is enabled
	  D: GPS operational

> #. Observe that the GPS starts to track the position based on the satellite data received as shown below:

Reviewer: ```suggestion
#. Observe that the GPS starts to track the position based on the satellite data received as follows:
```

	  D: Tracking SV 26:     used,     healthy
	  D: Tracking: 5 Using: 4 Unhealthy: 0
	  D: Seconds since last fix 30
	  D: Tracking SV 29:     used,     healthy
	  D: Tracking SV  5:     used,     healthy
	  D: Tracking SV  9:     used,     healthy
	  D: Tracking SV 21: not used,     healthy
	  D: Tracking SV 26:     used,     healthy
	  D: Tracking: 5 Using: 4 Unhealthy: 0
	  D: Seconds since last fix 31
	  D: Tracking SV 29:     used,     healthy
	  D: Tracking SV  5:     used,     healthy
	  D: Tracking SV  9:     used,     healthy
	  D: Tracking SV 21: not used,     healthy
	  D: Tracking SV 26:     used,     healthy
	  D: Tracking: 5 Using: 4 Unhealthy: 0
	  D: Seconds since last fix 32
	  D: Tracking SV 29:     used,     healthy
	  D: Tracking SV  5:     used,     healthy
	  D: Tracking SV  9:     used,     healthy
	  D: Tracking SV 21: not used,     healthy
	  D: Tracking SV 26:     used,     healthy
	  D: Tracking: 5 Using: 4 Unhealthy: 0
	  D: Seconds since last fix 33
	  D: PVT: Position fix
	  D: Stopping GPS

#. Observe that the current position of the device shows up on the GPS position map in nRF Cloud, after acquiring the GPS position fix.

>    The position details are also displayed on the terminal as shown below:

Reviewer: ```suggestion
   The position details are also displayed on the terminal as follows:
```

It is recommended to use LTE Power Saving Mode (PSM) and extended Discontinuous Reception (eDRX) mode to increase the allowed time of operation for GPS.
During the defined intervals of PSM and eDRX, LTE communication does not occur, and the GPS has full access to the radio resources.
In this sample, both PSM and eDRX are enabled by default.
You can enable or disable these features by using the :kconfig:`CONFIG_LTE_POWER_SAVING_MODE` and :kconfig:`CONFIG_LTE_EDRX_REQ` configuration options.

.. include:: /applications/asset_tracker/README.rst
   :start-after: external_antenna_note_start
   :end-before: external_antenna_note_end

User interface
**************

If :kconfig:`CONFIG_REST_DO_JITP` is enabled, pressing button 1 will request just-in-time provisioning with nRF Cloud.
This is useful when initially provisioning and associating a device on nRF Cloud.
You only need to do this once.

If :kconfig:`CONFIG_REST_ENABLE_LED` is enabled, LED1 indicates the state of the connection to nRF Cloud.

Configuration
*************
|config|

Configuration options
=====================

Check and configure the following configuration options for the sample:

> .. option:: CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode

Author: Fixing.

> .. option:: CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode
> 
> This configuration option enables or disables the LTE Power Saving Mode.

Reviewer: ```suggestion
.. kconfig:: CONFIG_LTE_POWER_SAVING_MODE

CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode
   This configuration option enables or disables the LTE Power Saving Mode.
```

Reviewer: If you decide **not** to add this option to the Kconfig file, please remove this and the reference to it on line 66.

Reviewer 2: I was going to comment about this, but seems that I forgot to do so. The name is a bit confusing, because for example CONFIG_LTE_EDRX_REQ is a configuration option for LTE LC, but CONFIG_LTE_POWER_SAVING_MODE is not. But the same naming is used in some other sample(s) as well. There's no Kconfig flag in LTE LC for PSM, but this needs to be in the application Kconfig and the application needs to call `lte_lc_psm_req()` function to enable PSM.

Author: I copied the readme from the agps sample and edited it, and did not notice that this kconfig was there but not in the code.  I will add it to the code to enable PSM mode, similar to how it is done in the agps sample.

> .. option:: CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode
> 
> This configuration option enables or disables the LTE Power Saving Mode.

Reviewer: ```suggestion
.. kconfig:: CONFIG_LTE_POWER_SAVING_MODE

CONFIG_LTE_POWER_SAVING_MODE - LTE Power Saving Mode
   This configuration option enables or disables the LTE Power Saving Mode.
```
